### PR TITLE
Regression: Unable to use Terms of Service with the core user plugin …

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #9390 .
#### Summary of Changes

added a check on `onUserBeforeSave`
#### Testing Instructions

> 1. install staging
> 2. go to the plugin manager
> 3. enable the plugin "User- Profile"
> 4. enable the option "Terms of Service:"
> 5. add a article (next option) for Terms of Service. (just one random article)
> 6. try to register from the frontend

### CODE REVIEW

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on fabda4779bb654858609b42fa58d7126c8c7ce63<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9405">issues.joomla.org/joomla-cms/9405</a>.</sub>

**Comment by brianteeman** [comment]
Dont know if programmatically it is good or bad but it works for me

**Comment by alikon** [comment]
i suspect shuold  be handled better on js side .... but i'm not that man ;)

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 4d88da852d2725a8da94f679076546ac623697cb<br /><br />:+1: Works good here! Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9405">issues.joomla.org/joomla-cms/9405</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 4d88da852d2725a8da94f679076546ac623697cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9405">issues.joomla.org/joomla-cms/9405</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9405">issues.joomla.org/joomla-cms/9405</a>.</sub>

**Comment by infograf768** [comment]
@wilsonge
This should go in as it solves a regression

File: plugins/user/profile/profile.php [hunk 1]
```diff
@@ -363,6 +363,11 @@ public function onUserBeforeSave($user, $isnew, $data)
 				throw new InvalidArgumentException(JText::_('PLG_USER_PROFILE_ERROR_INVALID_DOB'));
 			}
 		}
+		// Check that the tos are valid.
+		if (!($data['profile']['tos']))
```
Comment (joomdonation) [inline plugins/user/profile/profile.php]: Hmm

I haven't checked it carefully but this code might throws warning of the tos field is disabled ? Shouldn't it be 
```php
if (empty($data['profile']['tos']))
```

**Comment by alikon** [comment]
@joomdonation  can you check carefully #9439

**Comment by joomdonation** [comment]
@alikon The issue is still there. If you add this command 

```php
error_reporting(E_ALL);
```

before the command you added. And add this command after the command

```php
exit();
```

Then edit a user from backend, save it, there will be a notice displayed like this:

> Undefined index: tos in E:\www\beta3\plugins\user\profile\profile.php on line 369

**Comment by alikon** [comment]
@joomdonation thx for the review i'll add 
`
if (!empty($data["profile"]["tos"]))
{
  // code
}
`

**Comment by joomdonation** [comment]
Yes @alikon. That will prevent the notice.

**Comment by infograf768** [comment]
Can you make a patch?

**Comment by alikon** [comment]
I'll do this evening...

**Comment by joomdonation** [comment]
Actually, beside the notice, there is still bug in the current code. If tos is required, you won't be able to update your profile in frontend. 

We only need to validate tos when user register for an account, so I think we can change the current code to use this new code:

```php
$task = JFactory::getApplication()->input->getCmd('task');
$tosarticle = $this->params->get('register_tos_article');
$tosenabled = $this->params->get('register-require_tos', 0);

if ($task == 'register' && $tosarticle && $tosenabled == 2 && empty($data['profile']['tos']))
{
	throw new InvalidArgumentException(JText::_('PLG_USER_PROFILE_FIELD_TOS_DESC_SITE'));
}
```

**Comment by alikon** [comment]
@infograf768 @joomdonation 
please may i have a proof look #9458

>  If tos is required, you won't be able to update your profile in frontend.

i'm unable to replicate 
this is what i get for profile data  when visiting profile page [index.php/profile?task=profile.save](url)
`
'profile' => 
    array (size=11)
      'address1' => string '' (length=0)
      'address2' => string '' (length=0)
      'city' => string '' (length=0)
      'region' => string '' (length=0)
      'country' => string '' (length=0)
      'postal_code' => string '' (length=0)
      'phone' => string '' (length=0)
      'website' => boolean false
      'favoritebook' => string '' (length=0)
      'aboutme' => string '' (length=0)
      'dob' => string '' (length=0)

`

**Pull request merged**